### PR TITLE
fix(modules/asg): make tags_dest customizable with additional resources when needed

### PR DESCRIPTION
**Pull request by srihas619**
## Description

Fixing tags related bug while launching resources via ASG.
Addresses the issue https://github.com/PaloAltoNetworks/terraform-aws-swfw-modules/issues/29

### CODE REVIEW

**Comment by sebastianczech** [comment]
@srihas619 there is an error with `flatten` function:

```hcl
  │ Error: Too many function arguments
  │ 
  │   on main.tf line 31, in locals:
  │   31:   tags_dest                = flatten(["instance", "volume", "network-interface"], var.tags_dest_additional_resources)
  │     ├────────────────
  │     │ while calling flatten(list)
  │     │ var.tags_dest_additional_resources is a list of string
  │ 
  │ Function "flatten" expects only 1 argument(s).
  ╵}
```

In my opinion function `concat` should be used, not `flatten`.

Could you fix it, please ?

**Review by sebastianczech — CHANGES_REQUESTED** [review]
@srihas619 there is an error with `flatten` function:

```hcl
  │ Error: Too many function arguments
  │ 
  │   on main.tf line 31, in locals:
  │   31:   tags_dest                = flatten(["instance", "volume", "network-interface"], var.tags_dest_additional_resources)
  │     ├────────────────
  │     │ while calling flatten(list)
  │     │ var.tags_dest_additional_resources is a list of string
  │ 
  │ Function "flatten" expects only 1 argument(s).
  ╵}
```

In my opinion function `coalesce` should be used, not `flatten`, but then default value for variable `tags_dest_additional_resources` should be `null`, not `[]`. Moreover order should be different:

```hcl
coalesce(var.tags_dest_additional_resources, ["instance", "volume", "network-interface"])
```

Alternatively `concat` function can be used, if your target is to have values from both lists.

Could you fix it, please ?

**Comment by srihas619** [comment]
@sebastianczech  Thanks for the pointer. I would prefer to use `concat` in this case since it will allow more resources to be added when needed via variable.

**Comment by sebastianczech** [comment]
Now your changes work, but maybe we should keep it simpler? 

For example - do not use `concat` or `coalesce`, just remove `local.tags_dest` and use variable directly:

```hcl
  dynamic "tag_specifications" {
    for_each = toset(var.tags_dest)
    content {
      resource_type = tag_specifications.key
      tags          = local.tags_merged
    }
  }
```

Then variable will have default value - if somebody wants to define his/her custom list, it's dooable:

```hcl
variable "tags_dest" {
  description = "List of resources that need to be tagged when launched via autoscaling group"
  type        = list(string)
  default     = ["instance", "volume", "network-interface"]
}
```

**Comment by srihas619** [comment]
Thanks for validating changes.

Regarding your suggestion, I also gave a thought along these lines initially but then, I realized if we need to add more resources then we have to initialize this variable with defaults again and then append extra resources like `["instance", "volume", "network-interface", "<someother_resource1>","<someother_resource2>"]`

But I agree, it makes the implementation simpler, I will push an update.

**Comment by srihas619** [comment]
@sebastianczech Thanks for the quick approval and all the suggestions. Now, we wait for the merge and the release cut from it so we can unpin the `2.0.4` version.

File: modules/asg/variables.tf [hunk 1]
```diff
@@ -288,6 +288,12 @@ variable "security_group_ids" {
   default     = []
 }
 
+variable "tags_dest" {
```
Comment (michalbil) [inline modules/asg/variables.tf:291]:
```suggestion
variable "tag_specifications_targets" {
```
Can you please update the variable name to a more meaningful one?

Comment (srihas619) [reply modules/asg/variables.tf]: Thanks, updated it.

**Review by michalbil — APPROVED** [review]

**Pull request merged**